### DISCUSSION TITLE
Cumulative function input every unit of time?

### DISCUSSION THREAD

**Discussion by LukeFisherUKAEA** [idaholab/moose#19157] [Q&A Modules: General] [answered]
I have an initial condition function defined in a .csv file (e.g. a Gaussian) which is read in using the PiecewiseBilinear function.  I would like this function to be added to the system every time increment (e.g. 1 second), as a cumulative input to a variable.  For context, I am looking at modelling ion implantation and transport in a material, using a specific implantation profile.

Has anyone succeeded in making this work?  As far as I am aware there is no functionality to "activate" a new input every unit time from the PiecewiseBilinear (or other .csv input) function, so I have been looking to see if there are other kernels that achieve this.  Any advice would be much appreciated.

It may be better to define this profile as a function within MOOSE, rather than reading in and interpolating from a .csv - this would also be possible.

**Comment by GiudGiud** [comment, marked as answer]
Can you store the output of the function into an aux variable, then use an AccumulateAux to the accumulation in another variable ? https://mooseframework.inl.gov/source/auxkernels/AccumulateAux.html

**Reply by LukeFisherUKAEA** [reply]
This looks promising, thanks, although it does return an error stating that the variable I am assigning the accumulation to "does not exist in this system".  Additionally, do you know if there is a way to control the rate of accumulation, i.e. time between input "pulses", if you like?

**Reply by GiudGiud** [reply]
can you paste your auxvariables block and your auxkernels block here please?
AuxKernels can't act on anything else than auxiliary variables. 

Not natively. If you want to make a PR to add a function parameter to AccumulateAux for a rate that would be great.
If not, ~~you can also use the SelfAux to save an old copy of the variable~~, then use a ParsedAux to do the accumulation with a rate

**Reply by LukeFisherUKAEA** [reply]
That'll be the issue, as i've just discovered with some experimenting - that AuxKernels will only act on AuxVariables and not non-linear variables.  I'll see what I can do with the ParsedAux.

**Comment by hugary1995** [comment]
If you are dealing with a nonlinear variable, you need to add a source term. If you tell us the equation you are solving, we can advise you which kernel to use.
